PubSub/tests/System/PubSubTestCase.php

```diff
@@ -84,6 +89,13 @@ public static function topicAndSubscription(PubSubClient $client, array $topicCo
 
         return [$topic, $sub];
     }
+
+    public static function skipEmulatorTests($reason = null)
```

Comment:
How about putting this in `Google\Cloud\Core\Testing\System\SystemTestCase` so it can be used by other services? Only problem is I don't think you could use the static boolean, since it'll get toggled on/off globally for all the system test suites.

Maybe something like this:

```php
private $emulatorMap = [];

protected static function setUsingEmulator($testsuite, $enabled)
{
	 $emulatorMap[get_called_class()] = $enabled;
}

protected static function skipEmulatorTests($reason = null)
{
	if (isset($emulatorMap[get_called_class()] && $emulatorMap[get_called_class()]) {
		self::markTestSkipped($reason ?: "...");
	}
}
```

Reply:
We could remove `Google\Cloud\Spanner\Tests\System\SpannerTestCase::skipEmulatorTests()` and allow Spanner to use the new method as well.

Reply:
That would be handy, just need to slightly change the logic (match class namespace prefixes instead ot exact names, so we would need to set the flag only once, in `%LibName%TestCase`).

But we have emulators only for five libraries, and only three of them need special treatment. Do we really want a general solution here?

Reply:
> That would be handy, just need to slightly change the logic (match class namespace prefixes instead ot exact names, so we would need to set the flag only once, in %LibName%TestCase).

I thought about that, but I didn't think it was worth the extra effort. There aren't a large number of classes that would make optimization worthwhile, and if a system test were nested deeper in the file structure, the prefix check would either fall down or need to accomadate that.

I'm generally a fan of providing global helpers for this sort of thing. Even if it's only used by three clients it seems worthwhile given the minimal implementation.

Reply:
> I thought about that, but I didn't think it was worth the extra effort.

Not a big deal. Just a few more lines of code, a full scan of a small array in some cases.

> if a system test were nested deeper in the file structure, the prefix check would either fall down or need to accomadate that.

Well, all system test cases are placed in `Google\Cloud\%LibName%\Tests\System` or a nested namespace, so every test case name has `Google\Cloud\%LibName%\Tests\System\` prefix, I don't see a problem here. For more safety we can provide a way to choose one of two methods to set the flag: for exact class name and for a namespace prefix.

> Even if it's only used by three clients it seems worthwhile given the minimal implementation.

Agree.

Reply:
That sounds fine to me.